lgc/patch/PatchInOutImportExport.h

```diff
@@ -222,9 +225,10 @@ class PatchInOutImportExport : public Patch {
   llvm::GlobalVariable *m_lds; // Global variable to model LDS
   llvm::Value *m_threadId;     // Thread ID
 
-  std::vector<llvm::CallInst *> m_importCalls;                 // List of "call" instructions to import inputs
-  std::vector<llvm::CallInst *> m_exportCalls;                 // List of "call" instructions to export outputs
-  PipelineState *m_pipelineState = nullptr;                    // Pipeline state from PipelineStateWrapper pass
+  std::vector<llvm::CallInst *> m_importCalls;          // List of "call" instructions to import inputs
+  std::vector<llvm::CallInst *> m_exportCalls;          // List of "call" instructions to export outputs
+  std::map<unsigned, llvm::Value *[4]> m_attribExports; // Export info of vertex attributes
```

Comment:
I'd expect this to have a few values each, right? If that' the case, use SmallVector and SmallDenseSet.

Reply:
No, we must have 4 float values. So the array is fixed-length 4. If some components of <4 x float> are not specified, then we assign undef to them. Corresponds to this GLSL case in a typical VS: `layout(location = 0, component = 2) out vec2 v2`

Reply:
I meant replacing `m_importCalls` and `m_exportCalls` std::vectors with SmallVectors.
RE the map, C arrays don't follow value types semantics and have many gotchas. For example, you cannot return them from a function, they decay to pointers when passed to a function and the size is ignored. IMO, it's just too easy to misuse, especially with generic/templated code. `std::array<T, size>` has much more sane semantics, so I wanted to propose changing this to something like `llvm::SmallDenseMap<unsigned, std::array<llvm::Value *, 4>>`, if you don't care about the map order.

Reply:
OK. I will change it to std::array along with SmallDenseMap.